Bug in montecarlo absorption

gui-co opened the pull request:
**Description of work.**

This PR fixes a bug in MonteCarloAbsorption where the Y axis is always interpolated when the sparse instrument is used, even if it is not needed.

**To test:**

* Code review
* Use this script. Here the Monte Carlo simulation is done on sparse instrument but the number of bins is by default the same as the input. The interpolation is not needed. The script takes ~4min on main, less that 2 seconds with this PR (and ws1 and ws2 have same number of bins)

```python
from mantid.simpleapi import CreateSampleWorkspace, MonteCarloAbsorption

CreateSampleWorkspace(OutputWorkspace="ws1", XMin=0, XMax=5000, BinWidth=1, XUnit="Wavelength")
MonteCarloAbsorption(
    InputWorkspace="ws1",
    OutputWorkspace="ws2",
    SparseInstrument=True,
    Interpolation='CSpline',
    EventsPerPoint=1
    )
```

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
Interpolate only if needed

Commit:
Set calculated errors in output workspace

DannyHindson commented:
I think this is still a valid change. It probably didn't get picked up for review during the last release sprint because it didn't have a milestone set. @gui-co - are you happy to rebase this onto main? I can review it and test
Also I'm a bit surprised the CSpline interpolation is this slow. Your script takes even longer than 4 minutes on my machine using the main code. I'll profile that if I get chance and see what the cause is

DannyHindson commented:
I had a look at the slowness of the CSpline interpolation. It happens when the number of bins gets about a certain value. At a certain threshold the code stops using the fast analytical matrix inversion in void Matrix<T>::invertTridiagonal and reverts to a standard matrix inversion. This switch exists to stop an overflow in the sinh and cosh functions and I think you added this in https://github.com/mantidproject/mantid/pull/33896

I think the overflow can be avoided in a different way. The formula from the original paper that the code implements is:

![image](https://user-images.githubusercontent.com/56295817/189315474-97bf7b46-a874-4f19-bdd0-1649efce139f.png)

If the cosh and sinh functions are expanded out into:

![image](https://user-images.githubusercontent.com/56295817/189316185-61be2122-5f4b-4697-8ef4-0b1b527f2a7c.png)
![image](https://user-images.githubusercontent.com/56295817/189315742-51910ec4-ae13-4b76-b82a-33176273cbe4.png)


...then for large k I think there will be some cancellation leaving a formula that can be evaluated without overflow.

Not something for this PR but I might make this change at some point

gui-co commented:
Thanks for resurrecting this PR. I think I gave up because some system tests were failing and I didn't take the time to take a closer look.

I rebased on main and changed the base branch of the PR.

> I had a look at the slowness of the CSpline interpolation. It happens when the number of bins gets above a certain value. At a certain threshold the code stops using the fast analytical matrix inversion in void Matrix::invertTridiagonal and reverts to a standard matrix inversion. This switch exists to stop an overflow in the sinh and cosh functions and I think you added this in https://github.com/mantidproject/mantid/pull/33896

This is what I was trying right now (the example script takes ~5min without this PR and ~10 after).

DannyHindson commented:
I've had another think about this change. The spatial interpolation performed by MonteCarloAbsorption relies on the bins being the same on all histograms in the sparse workspace. If the input workspace to the MonteCarloAbsorption algorithm has different bins on some histograms, the algorithm creates a sparse workspace with equal spaced x Points by calculating a range based on the largest\smallest wavelength bin centre in the whole workspace and dividing this by the number of bins. For ISIS TOF diffraction instruments we find that the bins are the same on all histograms in the workspace that comes off the instrument because it's in TOF. When it's converted to wavelength though the bins are not the same on all histograms.

All of this means you have to do the wavelength interpolation following the spatial interpolation in a lot more situations than you might initially think in order to get an output workspace with the same wavelength values as the input workspace. So checking that `number of sparse bins`=`number of input bins` is not enough. You need to further check that all the bin centres in `spatiallyInterpHisto` and `targetHisto` are the same.

This will give the same speed up on the test workspace on this PR but may not speed up the absorption correction on a real workspace.

gui-co commented:
I am closing this, #34411 is properly fixing the issue.
